certbot/log.py

```diff
@@ -23,36 +24,82 @@
 
 
 def pre_arg_parse_setup():
-    """Ensures fatal exceptions are logged and reported to the user."""
-    sys.excepthook = functools.partial(except_hook, config=None)
+    """Setup logging before command line arguments are parsed.
+
+    Terminal logging is setup using
+    certbot.constants.QUIET_LOGGING_LEVEL so Certbot is as quiet as
+    possible. File logging is setup so that logging messages are
+    buffered in memory so they can either be written to a temporary
+    file before Certbot exits or to the normal logfiles once command
+    line arguments are parsed.
+
+    This function also sets logging.shutdown to be called on program
+    exit which automatically flushes logging handlers and sys.excepthook
+    to properly log/display fatal exceptions.
+
+    """
+    temp_handler = TempHandler()
+    temp_handler.setFormatter(logging.Formatter(FILE_FMT))
+    temp_handler.setLevel(logging.DEBUG)
+    memory_handler = MemoryHandler(temp_handler)
+
+    stream_handler = ColoredStreamHandler()
+    stream_handler.setFormatter(logging.Formatter(CLI_FMT))
+    stream_handler.setLevel(constants.QUIET_LOGGING_LEVEL)
+
+    root_logger = logging.getLogger()
+    root_logger.setLevel(logging.DEBUG)  # send all records to handlers
+    root_logger.addHandler(memory_handler)
+    root_logger.addHandler(stream_handler)
+
+    util.atexit_register(logging.shutdown)
```

Comment:
Will this still automatically flush the memory handler even if `shouldFlush` is set to always be `False`?